
# Discount factor effect and Q-Learning

## 1. Based on the probabilities on the arrows above, Model your MDP (transition probabilities and rewards) in the notebook. you can use the [s, a, s’] for each part or you can use your own way of defining the MDP. What should be seen are transition probabilities, rewards and possible actions

In [9]:
transition_probabilities = {
    's0': {'a0': [(0.7, 's0', 10), (0.3, 's1', 0)],
           'a1': [(1.0, 's0', 0)],
           'a2': [(0.8, 's0', 0), (0.2, 's1', 0)]},

    's1': {'a0': [(1.0, 's1', 0)],
           'a2': [(1.0, 's2', -50)]},

    's2': {'a1': [(0.1, 's2', 0), (0.1,'s1',0), (0.8, 's0', 40)]}
}
# Add reward to the transition_probabilities
# rewards

possible_actions = {'s0': ['a0', 'a1', 'a2'],
                    's1': ['a0', 'a2'],
                    's2': ['a1']}

states = ['s0', 's1', 's2']
actions = ['a0', 'a1', 'a2']

In [10]:
state_to_idx = {state: idx for idx, state in enumerate(states)}
action_to_idx = {action: idx for idx, action in enumerate(actions)}
print(f"state_to_idx: {state_to_idx}")
print(f"action_to_idx: {action_to_idx}")

state_to_idx: {'s0': 0, 's1': 1, 's2': 2}
action_to_idx: {'a0': 0, 'a1': 1, 'a2': 2}


In [11]:
for state in transition_probabilities:
    for action in transition_probabilities[state]:
        prob_sum = sum([prob for prob, _, _ in transition_probabilities[state][action]])
        if not np.isclose(prob_sum, 1.0):
            print(f"Warning: Probabilities for state {state} action {action} do not sum to 1: {prob_sum}")


In [12]:
# Q_values = np.full((3, 3), -np.inf)  # -np.inf for impossible actions
# for state, actions in enumerate(possible_actions):
#     Q_values[state, actions] = 0.0  # for all possible actions
import numpy as np
Q_values = np.full((len(states), len(actions)), -np.inf)  # 3 states, 3 actions
for state in possible_actions:
    for action in possible_actions[state]:
        Q_values[state_to_idx[state], action_to_idx[action]] = 0.0

In [13]:
print("Transition Probabilities are valid.")
print(f"state_to_idx: {state_to_idx}")
print(f"action_to_idx: {action_to_idx}")
print("Initialized Q_values:")
print(Q_values)

Transition Probabilities are valid.
state_to_idx: {'s0': 0, 's1': 1, 's2': 2}
action_to_idx: {'a0': 0, 'a1': 1, 'a2': 2}
Initialized Q_values:
[[  0.   0.   0.]
 [  0. -inf   0.]
 [-inf   0. -inf]]


### **ANSWER Q7.1**:
* In transition probabilities, I defined `transition_probabilities` is a dictionaries of keys are initial states (s0, s1, s2) and values are dictionaries including keys are actions (a0, a1, a2) which are depended on the states (each states have specific actions to take), and values are list in form of `[probabilities, next_state, reward]`

* The `possible_actions` include specific actions corresponding to each states that can be taken


## 2. Take your discount factor to be 0.9. Perform Q-learning and report the Q-values for each (state, action) pair. Based on that, what is the optimal policy?

In [14]:
gamma = 0.9  # the discount factor

#gamma = 0.95
for iteration in range(50):
    # --- fill here (perform a DP approach for filling up your Q-table (repeat the process by stting gamma to be 0.95 )
    new_Q_values = np.copy(Q_values)
    for state in states:
      state_idx = state_to_idx[state]
      for action in possible_actions[state]:
        action_idx = action_to_idx[action]
        value_sum = 0
        for prob, next_state, reward in transition_probabilities[state][action]:
          next_state_idx = state_to_idx[next_state]
          best_next_action = np.argmax(new_Q_values[next_state_idx])
          value_sum += prob * (reward + gamma * new_Q_values[next_state_idx][best_next_action])

        new_Q_values[state_idx, action_idx] = value_sum
    Q_values = new_Q_values

In [15]:
print("Q-values with gamma = 0.9:")
print(Q_values)

Q-values with gamma = 0.9:
[[18.91891892 17.02702703 13.62162162]
 [ 0.                -inf -4.87971488]
 [       -inf 50.13365013        -inf]]


In [16]:
# Q_values.argmax(axis=1)  # optimal action for each state
optimal_policy = [actions[np.argmax(Q_values[state_to_idx[state]])] for state in states]
print("Optimal policy with gamma = 0.9:")
print(optimal_policy)

Optimal policy with gamma = 0.9:
['a0', 'a0', 'a1']


### **ANSWER Q7.2**:
The highest Q-value for `s0` is for action `a0`, for `s1` is action `a0` and for `s2` is action `a1` with Q value of 50.13.

Thus the optimal policy is

* From state s0, the optimal action is a0.
* From state s1, the optimal action is a0.
* From state s2, the optimal action is a1.




## 3. Perform the same procedure but this time with a discount factor of 0.95. Did your optimal policy change? Explain your results.



In [17]:
# you can use the same code as above
Q_values = np.full((len(states), len(actions)), -np.inf)  # Reset Q-values
for state in possible_actions:
    for action in possible_actions[state]:
        Q_values[state_to_idx[state], action_to_idx[action]] = 0.0


gamma = 0.95  # New discount factor
for iteration in range(50):
    # --- fill here (perform a DP approach for filling up your Q-table (repeat the process by stting gamma to be 0.95 )
    new_Q_values = np.copy(Q_values)
    for state in states:
      state_idx = state_to_idx[state]
      for action in possible_actions[state]:
        action_idx = action_to_idx[action]
        value_sum = 0
        for prob, next_state, reward in transition_probabilities[state][action]:
          next_state_idx = state_to_idx[next_state]
          best_next_action = np.argmax(new_Q_values[next_state_idx])
          value_sum += prob * (reward + gamma * new_Q_values[next_state_idx][best_next_action])

        new_Q_values[state_idx, action_idx] = value_sum
    Q_values = new_Q_values

In [18]:
# Print Q-values
print("Q-values with gamma = 0.95:")
print(Q_values)

Q-values with gamma = 0.95:
[[21.79615996 20.70635196 16.76923123]
 [ 1.02074831        -inf  1.08097586]
 [       -inf 53.77587186        -inf]]


In [19]:
optimal_policy = [actions[np.argmax(Q_values[state_to_idx[state]])] for state in states]
print("Optimal policy with gamma = 0.95:")
print(optimal_policy)

Optimal policy with gamma = 0.95:
['a0', 'a2', 'a1']


### **ANSWER Q7.3**:

The highest Q-value for `s0` is for action `a0`, for `s1` is action `a2` and for `s2`, action `a1` has a Q-value of 53.78.

Now the optimal policy is:
* From state s0, the optimal action is a0.
* From state s1, the optimal action is a2.
* From state s2, the optimal action is a1.



### **Detail analysis and result explanation**

For discount factor `gamma = 0.95`, the optimal policy changes, especially in state 2 from taking action `a0` (`gamma = 0.9`) to action `a2`. The changes is due to higher discount factors enphasize more importance on future reward. This indicates a consideration of future rewards despite the immediate cost of -50. This is likely due to reaching s2 has a potential high future reward (the transition from s2 back to s0 with a1).


### **ANSWER Q7.4**:

Discount factor determined the importance of future rewards in the calculation of total expected reward. It is used in Q-learning update formula:

$$
Q(s, a) = \sum_{s'} P(s' \mid s, a) \left[ R(s, a, s') + \gamma \max_{a'} Q(s', a') \right]
$$

The discount factor $\gamma$ is a value between 0 and 1:
- When $\gamma$ is close to 1, future rewards are highly valued, encouraging long-term strategy.
- When $\gamma$ is close to 0, immediate rewards are prioritized, focusing on short-term gains.

if we modify the discount factor, the optimal policy can changed based on the value of $\gamma$ as we demonstrated above

